### PR TITLE
Compile gtsam python for windows

### PR DESCRIPTION
**Pull request by talregev**
Compile gtsam python for windows

I manage to compile and run python tests on windows.
Some of the python tests are fails on windows (16 of 243). 
Please take a look on my PR.

@dellaert @varunagrawal @ProfFan 

Waiting for merge:
- [x] #1704

### CODE REVIEW

**Comment by talregev** [comment]
I will disable the running of tests for windows. that you have a clean CI.

**Comment by ProfFan** [comment]
Thank you! Could you post which tests fail here?

**Comment by talregev** [comment]
I am post here the errors, also I am put a link for the run:
https://github.com/borglab/gtsam/actions/runs/7135793022/job/19433202623
```
======================================================================
ERROR: test_dogleg (test_NonlinearOptimizer.TestScenario.test_dogleg)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_gauss_newton (test_NonlinearOptimizer.TestScenario.test_gauss_newton)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_gnc_params (test_NonlinearOptimizer.TestScenario.test_gnc_params)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_graduated_non_convexity (test_NonlinearOptimizer.TestScenario.test_graduated_non_convexity)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_iteration_hook (test_NonlinearOptimizer.TestScenario.test_iteration_hook)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_levenberg_marquardt (test_NonlinearOptimizer.TestScenario.test_levenberg_marquardt)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_levenberg_marquardt_pcg (test_NonlinearOptimizer.TestScenario.test_levenberg_marquardt_pcg)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_NonlinearOptimizer.py", line 43, in setUp
    self.assertEqual(0.0, self.fg.error(self.optimal_values), 0.0)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_SFMExample (test_SFMExample.TestSFMExample.test_SFMExample)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_SFMExample.py", line 63, in test_SFMExample
    optimizer = gtsam.LevenbergMarquardtOptimizer(graph, initialEstimate)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "p0", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>

======================================================================
ERROR: test_StereoVOExample (test_StereoVOExample.TestStereoVOExample.test_StereoVOExample)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_StereoVOExample.py", line 71, in test_StereoVOExample
    optimizer = gtsam.LevenbergMarquardtOptimizer(graph, initialEstimate)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "l1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>

======================================================================
ERROR: test_localToWorld (test_Utilities.TestUtilites.test_localToWorld)
Test localToWorld.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_Utilities.py", line 182, in test_localToWorld
    np.testing.assert_allclose(world.atPoint2(0), expected_point2)
                               ^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to at the key "0", which does not exist in the Values.

======================================================================
ERROR: test_perturbPoint2 (test_Utilities.TestUtilites.test_perturbPoint2)
Test perturbPoint2.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_Utilities.py", line 113, in test_perturbPoint2
    not np.allclose(values.atPoint2(1), gtsam.Point2(1, 1)))
                    ^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_perturbPoint3 (test_Utilities.TestUtilites.test_perturbPoint3)
Test perturbPoint3.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_Utilities.py", line 130, in test_perturbPoint3
    self.assertTrue(not np.allclose(values.atPoint3(1), point3))
                                    ^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>

======================================================================
ERROR: test_values (test_Values.TestValues.test_values)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_Values.py", line 62, in test_values
    self.gtsamAssertEquals(values.atPoint2(0), Point2(0,0), tol)
                           ^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "0", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,2,1,0,2,1>

======================================================================
ERROR: test_VisualISAMExample (test_VisualISAMExample.TestVisualISAMExample.test_VisualISAMExample)
Test to see if ISAM works as expected for a simple visual SLAM example.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_VisualISAMExample.py", line 48, in test_VisualISAMExample
    isam, result, nextPose = visual_isam.initialize(
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\gtsam\gtsam\build\python\gtsam\utils\visual_isam.py", line 67, in initialize
    batchOptimizer = gtsam.LevenbergMarquardtOptimizer(newFactors,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "l0", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>

======================================================================
ERROR: test_isam2_error (test_VisualISAMExample.TestVisualISAMExample.test_isam2_error)
Test for isam2 error() method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_VisualISAMExample.py", line 67, in test_isam2_error
    isam, result, nextPose = visual_isam.initialize(
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\gtsam\gtsam\build\python\gtsam\utils\visual_isam.py", line 67, in initialize
    batchOptimizer = gtsam.LevenbergMarquardtOptimizer(newFactors,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "l0", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>

======================================================================
ERROR: test_isam2_update (test_VisualISAMExample.TestVisualISAMExample.test_isam2_update)
Test for full version of ISAM2::update method
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\gtsam\gtsam\build\python\gtsam\tests\test_VisualISAMExample.py", line 94, in test_isam2_update
    isam, result, nextPose = visual_isam.initialize(
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\gtsam\gtsam\build\python\gtsam\utils\visual_isam.py", line 67, in initialize
    batchOptimizer = gtsam.LevenbergMarquardtOptimizer(newFactors,
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Attempting to retrieve value with key "l0", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>

----------------------------------------------------------------------

Ran 243 tests in 0.543s



FAILED (errors=16, skipped=3)

```

**Comment by ProfFan** [comment]
It seems to be an RTTI issue, which actually makes the Python wrappers very cumbersome to use....

**Comment by talregev** [comment]
> It seems to be an RTTI issue, which actually makes the Python wrappers very cumbersome to use....

I am not sure what are you suggesting.

**Comment by talregev** [comment]
I am not sure what do you mean by RTTI issue. 
I did some debug.
The error we get from here:
https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/Values.cpp#L297

That was thrown from here:
https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/Values-inl.h#L141

Maybe is msvc bug? 
The flow of the code is like that. start from here:
https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/Values-inl.h#L213
For the handle class there is 2 option. 
General case:
https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/Values-inl.h#L134
Eigen::Matrix case (template specialization):
https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/Values-inl.h#L193

On windows it choose the general case.
On Linux it choose the  Eigen::Matrix case.

Also for linux:
```
storedTypeName: gtsam::GenericValue<Eigen::Matrix<double, -1, 1, 0, -1, 1> >
requestedTypeName: Eigen::Matrix<double, -1, 1, 0, -1, 1>
```

**Comment by ProfFan** [comment]
Yes, that is an RTTI issue, because `dynamic_cast` failed in the previous step

**Comment by talregev** [comment]
I suspect it before the dynamic cast. It from template specialization.
Because it choose the incorrect (or different from Linux) case for the template.

**Comment by ProfFan** [comment]
I fixed this problem before, don't know why it's coming up again. Culprit is likely `type_info` being inconsistent between different dynamic libraries. #218

**Comment by ProfFan** [comment]
Also https://github.com/borglab/GTDynamics/issues/248

**Comment by talregev** [comment]
@ProfFan I added a workaround for windows template specialization. 
I test it local. (I copy the gtsam code, and just change the template specialization):
https://github.com/borglab/gtsam/pull/1685/files#diff-2d6a3f78e983bc926a126f73463c25805ef90f9e838f91af83192f68b48353b9R200
Now it compile and run correctly.
It not a problem of `dynamic_cast` , nor a problem of `type_info`
It a problem of msbuild template substitute. I think it msbuild compiler bug. (when I compile this code on gcc, it create ambiguous error). msbuild not substitute correctly the template. 
You can look at the code and see for yourself.

~Now I have a new problem, from latest merge #1572:~
~It not found the `cephes_igami`~
```
nonlinear.cpp.obj : error LNK2001: unresolved external symbol cephes_igami
```

@varunagrawal @ProfFan  ~Can you help me to add this symbol (or give an hint) to windows python?~
Thank you in advanced, and Happy new year. 

-----------------
Edit: 
I added workaround for export `cephes_igami` for windows.

**Comment by ProfFan** [comment]
<del>What is cephes_igami?</del> nvm

**Comment by talregev** [comment]
> What is cephes_igami?

`cephes_igami` is a define from 3rdparty cephes that was added here: #1572
https://github.com/borglab/gtsam/blob/develop/gtsam/3rdparty/cephes/cephes/cephes_names.h#L45
In my workaround, I export the `igami` function. then it also export the `cephes_igami` because of that define.

File: gtsam/3rdparty/cephes/dllexport.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/* ----------------------------------------------------------------------------
```
Comment (ProfFan) [inline gtsam/3rdparty/cephes/dllexport.h]: Copy-paste? if this is just an Windows issue just add the required export for Windows

File: gtsam/nonlinear/Values-inl.h [hunk 2]
```diff
@@ -197,6 +197,59 @@ namespace gtsam {
      }
    };
 
+#ifdef _WIN32
+   // Handle dynamic matrices
+   template <int M, int N>
+   struct handle_matrix<Eigen::Matrix<double, M, N, 0, M, N>, true> {
```
Comment (ProfFan) [inline gtsam/nonlinear/Values-inl.h:205]: I don't really understand why this is not in the right case, it may indicate a bigger problem which is the default case being different in different compilation units...

**Review by ProfFan — COMMENTED** [review]
Thank you for the contributions! I'll leave the rest of review to @varunagrawal who is better suited as Windows reviewer. Also CC @mattking-smith @jlblancoc who may be more familiar with Windows.

File: gtsam/3rdparty/cephes/dllexport.h (hunk 1 above)
Comment (talregev) [reply gtsam/3rdparty/cephes/dllexport.h]: Copy paste. I didn't want to change something that already working.

File: gtsam/nonlinear/Values-inl.h (hunk 2 above)
Comment (talregev) [reply gtsam/nonlinear/Values-inl.h:205]: I don't understand it either. My suspicion is this an msvc compiler bug.
On gcc compiler on this code, I get ambiguous error as I expected.

Comment (ProfFan) [reply gtsam/nonlinear/Values-inl.h:205]: What does the full MSVC substitution look like?

File: gtsam/3rdparty/cephes/dllexport.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/3rdparty/cephes/dllexport.h]: @talregev I'm making a Cephes related PR, I will update it to include the proper `extern` definitions.

File: gtsam/nonlinear/Values-inl.h (hunk 2 above)
Comment (talregev) [reply gtsam/nonlinear/Values-inl.h:205]: I don't know. I just guess this one according to the input.

File: gtsam/3rdparty/cephes/dllexport.h (hunk 1 above)
Comment (talregev) [reply gtsam/3rdparty/cephes/dllexport.h]: @varunagrawal Thank you! I will rebase your PR after it will be on dev.
Can you review the rest of the PR?

File: .github/scripts/python.sh [hunk 3]
```diff
@@ -65,14 +65,13 @@ function build()
...
 
-  $PYTHON -m pip install --user build/python
+  cmake --build build --target python-install
 }
 
 function test()
 {
-  cd $GITHUB_WORKSPACE/python/gtsam/tests
-  $PYTHON -m unittest discover -v
-  cd $GITHUB_WORKSPACE
```
Comment (varunagrawal) [inline .github/scripts/python.sh:75]: Please undo these changes. One reason we keep them here is so that others can copy-paste them when trying to run tests, and these changes are tangential to the PR.

File: .github/workflows/build-python.yml [hunk 4]
```diff
@@ -18,9 +18,11 @@ jobs:
       CTEST_PARALLEL_LEVEL: 2
       CMAKE_BUILD_TYPE: ${{ matrix.build_type }}
       PYTHON_VERSION: ${{ matrix.python_version }}
+      BOOST_VERSION: 1.72.0
+      BOOST_EXE: boost_1_72_0-msvc-14.2      
 
     strategy:
-      fail-fast: true
+      fail-fast: false
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml]: What was the need for this? We would like the CI to fail immediately on errors, whereas full error reporting should ideally be done locally.

File: .github/workflows/build-python.yml [hunk 5]
```diff
@@ -97,29 +104,68 @@ jobs:
...
+          $BOOST_PATH = "C:\hostedtoolcache\windows\Boost\$env:BOOST_VERSION\x86_64"
+
+          # Use the prebuilt binary for Windows
+          $Url = "https://sourceforge.net/projects/boost/files/boost-binaries/$env:BOOST_VERSION/$env:BOOST_EXE-${{matrix.platform}}.exe"
+          (New-Object System.Net.WebClient).DownloadFile($Url, "$env:TEMP\boost.exe")
+          Start-Process -Wait -FilePath "$env:TEMP\boost.exe" "/SILENT","/SP-","/SUPPRESSMSGBOXES","/DIR=$BOOST_PATH"
+
+          # Set the BOOST_ROOT variable
+          echo "BOOST_ROOT=$BOOST_PATH" >> $env:GITHUB_ENV
+
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml:141]: This all seems tangential to the problem this PR is trying to address.

File: .github/workflows/build-python.yml [hunk 6]
```diff
@@ -97,29 +104,68 @@ jobs:
...
       - name: Set GTSAM_WITH_TBB Flag
         if: matrix.flag == 'tbb'
         run: |
           echo "GTSAM_WITH_TBB=ON" >> $GITHUB_ENV
           echo "GTSAM Uses TBB"
 
-      - name: Set Swap Space
+      - name: Set Swap Space (Linux)
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml:148]: Tangential again

File: .github/workflows/build-python.yml [hunk 7]
```diff
@@ -97,29 +104,68 @@ jobs:
...
         if: runner.os == 'Linux'
         uses: pierotofy/set-swap-space@master
         with:
           swap-size-gb: 6
 
-      - name: Install System Dependencies
+      - name: Install System Dependencies (Linux, macOS)
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml:154]: Tangential again

File: .github/workflows/build-python.yml [hunk 8]
```diff
@@ -97,29 +104,68 @@ jobs:
...
+        if: runner.os != 'Windows'
         run: |
           bash .github/scripts/python.sh -d
 
       - name: Install Python Dependencies
+        shell: bash
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml:160]: Tangential again

File: .github/workflows/build-python.yml [hunk 9]
```diff
@@ -97,29 +104,68 @@ jobs:
...
         run: python$PYTHON_VERSION -m pip install -r python/dev_requirements.txt
 
       - name: Build
+        shell: bash
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml:164]: Tangential again

File: .github/workflows/build-python.yml [hunk 10]
```diff
@@ -97,29 +104,68 @@ jobs:
...
         run: |
           bash .github/scripts/python.sh -b
 
       - name: Test
+        shell: bash
```
Comment (varunagrawal) [inline .github/workflows/build-python.yml:169]: Tangential again

File: gtsam/3rdparty/cephes/CMakeLists.txt [hunk 11]
```diff
@@ -8,6 +8,7 @@ project(
 
 set(CEPHES_HEADER_FILES
     cephes.h
+    dllexport.h
```
Comment (varunagrawal) [inline gtsam/3rdparty/cephes/CMakeLists.txt]: You can undo this and other related changes (aka `git revert`) and then rebase the latest develop after #1704 is merged.

File: .github/workflows/build-python.yml (hunk 4 above)
Comment (talregev) [reply .github/workflows/build-python.yml]: > What was the need for this? We would like the CI to fail immediately on errors, whereas full error reporting should ideally be done locally.

On this one, we run multiples os system, with different configuration, that one configure is not connected to another.
On my develop on windows, I did false, to see if the other configs are ok, even the windows was fail. when the flag is true, it really hard for dev.
I already give you that if you push another commit to the PR, all the previous job will terminate automatic. 
This is the ideal for develop.

File: .github/workflows/build-python.yml (hunk 5 above)
Comment (talregev) [reply .github/workflows/build-python.yml:141]: No, it necessary to have boost for compiling with gtsam python. You cannot delete these lines. I think it because using the boost serialize (the same happen for matlab).

File: .github/workflows/build-python.yml (hunk 6 above)
Comment (talregev) [reply .github/workflows/build-python.yml:148]: This is only change the name. I didn't create this line. It better to know that these only happen for linux.
Do you want me to remove these lines that you are written?

File: .github/workflows/build-python.yml (hunk 8 above)
Comment (talregev) [reply .github/workflows/build-python.yml:160]: This change must happen, because I an running shell scrip on windows, so I need to tell it that it will bash.

File: .github/workflows/build-python.yml (hunk 9 above)
Comment (talregev) [reply .github/workflows/build-python.yml:164]: This change must happen, because I an running shell scrip on windows, so I need to tell it that it will bash.

File: .github/workflows/build-python.yml (hunk 10 above)
Comment (talregev) [reply .github/workflows/build-python.yml:169]: This change must happen, because I an running shell scrip on windows, so I need to tell it that it will bash.

File: gtsam/3rdparty/cephes/CMakeLists.txt (hunk 11 above)
Comment (talregev) [reply gtsam/3rdparty/cephes/CMakeLists.txt]: I will rebase your PR and then undo / resolve the conflicts.

File: .github/workflows/build-python.yml (hunk 7 above)
Comment (talregev) [reply .github/workflows/build-python.yml:154]: Name change for better understand that it only happen on linux and macos.

**Comment by talregev** [comment]
Except the name changes, all the changes here are must and necessary to solve the problem of compiling windows gtsam python.

File: .github/scripts/python.sh (hunk 3 above)
Comment (talregev) [reply .github/scripts/python.sh:75]: I will revert. 
people also can copy the cmake commands. It will work on local the same.

File: .github/workflows/build-python.yml (hunk 4 above)
Comment (talregev) [reply .github/workflows/build-python.yml]: Even when false, the CI will fail when only one job is failing while other will continue to run. When true, you cannot see the other job results.
You assuming that people dev on local, then publish the PR. but the truth is people develops and relay on CI for dev like me.

File: .github/scripts/python.sh (hunk 3 above)
Comment (talregev) [reply .github/scripts/python.sh:75]: Also people cannot copy past these line with GITHUB_WORKSPACE environment variable, they need to modify it.
On my lines, they don't need to change anything, plus it bonus that it using the python that was discovered in cmake, ratter than the default python that they have, so in some cases it not the same.

File: .github/workflows/build-python.yml (hunk 10 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml:169]: Run it via a different script then instead of using the workflow file? Or is there a reason you're doing it this way?

File: .github/workflows/build-python.yml (hunk 7 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml:154]: It is distracting from the intended changes and making it hard to review. If you want me to review this PR, please accept my comments as I have limited time.

File: .github/workflows/build-python.yml (hunk 5 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml:141]: The current CI passes without problems so I don't see how this makes any difference. Can you please explain in detail?

File: .github/scripts/python.sh (hunk 3 above)
Comment (varunagrawal) [reply .github/scripts/python.sh:75]: When I say copy-paste, I mean with modifications. The original code was also meant to be instructional since people are familiar with how to run Python's unittest module and may not have the same experience with CMake.

Again, if you wish to receive reviews from me, please accept my comments. I only request changes when needed since we have put considerable thought into these things in the past.

File: .github/workflows/build-python.yml (hunk 4 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml]: Sorry but we have limited CI resources since we don't get paid for this project. You should not be using our CI for your testing purposes and bottlenecking other PRs.

File: .github/workflows/build-python.yml (hunk 5 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml:141]: Nvm I understand this now.

File: .github/workflows/build-python.yml (hunk 6 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml:148]: Actually on second thought I like what you've done here, let's keep these changes. :)

File: gtsam/nonlinear/Values-inl.h (hunk 2 above)
Comment (varunagrawal) [reply gtsam/nonlinear/Values-inl.h:205]: Can you please add a comment here about what this is doing and why is it needed? The comment can state that this is due to a possible MSVC bug.

File: .github/workflows/build-python.yml (hunk 5 above)
Comment (talregev) [reply .github/workflows/build-python.yml:141]: Resolve these comments.

File: .github/workflows/build-python.yml (hunk 4 above)
Comment (talregev) [reply .github/workflows/build-python.yml]: I am not using gtsam CI, I am using my CI for dev.

File: .github/scripts/python.sh (hunk 3 above)
Comment (talregev) [reply .github/scripts/python.sh:75]: I already undo these changes.

Comment (talregev) [reply .github/scripts/python.sh:75]: This should be outdated.

File: .github/workflows/build-python.yml (hunk 10 above)
Comment (talregev) [reply .github/workflows/build-python.yml:169]: The same reason why you are running one scrip for Linux and macos code. You have one script that already working, less code to review and debug and less code duplication. Changes can be in one place, less changes for get error and bug. (instead change need to do in multiple places). 
If I will do another script for windows, it will be code duplication, more code that you will need to review with your limited time, more time for me to implement and debug, Instead run script that already working.

File: gtsam/nonlinear/Values-inl.h (hunk 2 above)
Comment (talregev) [reply gtsam/nonlinear/Values-inl.h:205]: I added

**Comment by talregev** [comment]
This branch is ready for review.

File: .github/workflows/build-python.yml (hunk 10 above)
Comment (talregev) [reply .github/workflows/build-python.yml:169]: @varunagrawal Can we have the same script for all os?

File: .github/workflows/build-python.yml (hunk 4 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml]: Then please undo this change before pushing here. You can make the change for your branch, test on your CI and then undo the change before pushing it here. :)

**Review by varunagrawal — COMMENTED** [review]
Asking for one small change and then I can approve this PR.

**Review by varunagrawal — APPROVED** [review]
